File: app/src/main/java/uiapp/widget/TitleBar.java
package uiapp.widget;

import android.app.Activity;
import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.drawable.Drawable;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.FragmentActivity;
import androidx.fragment.app.FragmentManager;

import uiapp.R;

/**
 * 标题栏
 * <p>默认注册了返回的监听, 关闭 Activity 或 Fragment
 */
public class TitleBar extends FrameLayout {
    private TextView titleView;
    private ImageView leftView;
    private ImageView rightView;
    private TextView rightText;
    private View lineView;

    public TitleBar(Context context) {
        this(context, null, 0);
    }

    public TitleBar(@NonNull Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public TitleBar(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        this(context, attrs, defStyleAttr, 0);
    }

    public TitleBar(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        LayoutInflater.from(context).inflate(R.layout.titlebar, this, true);

        initialize(context, attrs, defStyleAttr, defStyleRes);

        setLeftClickListener(v -> {
            if (getContext() instanceof FragmentActivity) {
                FragmentActivity activity = (FragmentActivity) getContext();
                FragmentManager manager = activity.getSupportFragmentManager();
                if (manager.getBackStackEntryCount() > 0) {
                    manager.popBackStack();
                } else {
                    activity.finish();
                }
            } else if (getContext() instanceof Activity) {
                ((Activity) getContext()).finish();
            }
        });
    }

    private void initialize(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        titleView = findViewById(R.id.title);
        leftView = findViewById(R.id.left_icon);
        rightView = findViewById(R.id.right_icon);
        rightText = findViewById(R.id.right_text);
        lineView = findViewById(R.id.line);

        final TypedArray a = context.obtainStyledAttributes(attrs, R.styleable.TitleBar, defStyleAttr, defStyleRes);
        String titleText = a.getString(R.styleable.TitleBar_title);
        Drawable leftIcon = a.getDrawable(R.styleable.TitleBar_leftIcon);
        Drawable rightIcon = a.getDrawable(R.styleable.TitleBar_rightIcon);
        String rightText = a.getString(R.styleable.TitleBar_rightText);
        boolean lineVisible = a.getBoolean(R.styleable.TitleBar_lineVisible, true);
        boolean leftVisible = a.getBoolean(R.styleable.TitleBar_leftVisible, true);
        a.recycle();

        if (titleText != null) {
            setTitle(titleText);
        }
        if (rightText != null) {
            setRightText(rightText);
        }
        if (rightIcon != null) {
            setRightView(rightIcon);
        }
        if (leftIcon != null) {
            setRightView(leftIcon);
        }
        leftView.setVisibility(leftVisible ? VISIBLE : GONE);
        lineView.setVisibility(lineVisible ? VISIBLE : GONE);
    }

    public void setLeftClickListener(OnClickListener listener) {
        this.leftView.setOnClickListener(listener);
    }

    public void setRightClickListener(OnClickListener listener) {
        this.rightView.setOnClickListener(listener);
        this.rightText.setOnClickListener(listener);
    }

    public TextView getTitleView() {
        return titleView;
    }

    public ImageView getLeftView() {
        return leftView;
    }

    public ImageView getRightView() {
        return rightView;
    }

    public TextView getRightText() {
        return rightText;
    }

    public View getLineView() {
        return lineView;
    }

    public void setTitle(CharSequence title) {
        titleView.setText(title);
    }

    public void setLeftView(Drawable drawable) {
        leftView.setImageDrawable(drawable);
    }

    public void setRightView(Drawable drawable) {
        rightView.setImageDrawable(drawable);
    }

    public void setRightText(CharSequence text) {
        rightText.setText(text);
    }
}

File: app/src/main/java/uiapp/widget/TitleBar.kt
package uiapp.widget

import android.app.Activity
import android.content.Context
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.TextView
import androidx.fragment.app.FragmentActivity
import uiapp.R

/**
 * 标题栏
 *
 * 默认注册了返回的监听, 关闭 Activity 或 Fragment
 */
class TitleBar @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet?,
    defStyleAttr: Int = 0,
    defStyleRes: Int = 0
) : FrameLayout(context, attrs, defStyleAttr, defStyleRes) {

    @JvmField
    val title: TextView

    @JvmField
    val left: ImageView

    @JvmField
    val right: TextView

    @JvmField
    val line: View


    init {
        LayoutInflater.from(context).inflate(R.layout.titlebar, this, true)
        title = findViewById(R.id.title)
        left = findViewById(R.id.left)
        right = findViewById(R.id.right)
        line = findViewById(R.id.line)

        initialize(context, attrs, defStyleAttr, defStyleRes)

        left.setOnClickListener {
            if (getContext() is FragmentActivity) {
                val activity = getContext() as FragmentActivity
                val manager = activity.supportFragmentManager
                if (manager.backStackEntryCount > 0) {
                    manager.popBackStack()
                } else {
                    activity.finish()
                }
            } else if (getContext() is Activity) {
                val activity = getContext() as FragmentActivity
                activity.finish()
            }
        }
    }

    private fun initialize(
        context: Context,
        attrs: AttributeSet?,
        defStyleAttr: Int,
        defStyleRes: Int
    ) {
        val a =
            context.obtainStyledAttributes(attrs, R.styleable.TitleBar, defStyleAttr, defStyleRes)
        val titleText = a.getString(R.styleable.TitleBar_title)
        val leftIcon = a.getDrawable(R.styleable.TitleBar_leftIcon)
        val rightIcon = a.getDrawable(R.styleable.TitleBar_rightIcon)
        val rightText = a.getString(R.styleable.TitleBar_rightText)
        val leftVisible = a.getBoolean(R.styleable.TitleBar_leftVisible, true)
        val lineVisible = a.getBoolean(R.styleable.TitleBar_lineVisible, true)
        a.recycle()

        if (titleText != null) {
            title.text = titleText
        }
        if (leftIcon != null) {
            left.setImageDrawable(leftIcon)
        }
        if (rightText != null) {
            right.text = rightText
            right.visibility = VISIBLE
        }
        if (rightIcon != null) {
            right.setCompoundDrawablesRelative(rightIcon, null, null, null)
            right.visibility = VISIBLE
        }
        left.visibility = if (leftVisible) VISIBLE else GONE
        line.visibility = if (lineVisible) VISIBLE else GONE
    }

}